docs/reference/src/main/asciidoc/contexts.asciidoc

```diff
@@ -250,3 +250,166 @@ conversation context, as can the current conversation.
 
 NOTE: Weld's conversations are not assigned ids until they become
 non-transient.
+
+=== Propagating built-in contexts
+
+By context propagation we understand a scenario in which you want to capture a collection of contextual instances bound to certain context in one thread and provide them as context state in another thread.
+Starting with Weld 3.1.0.Final, this kind of context propagation is possible.
+
+NOTE: Context propagation comes with some additional requirements on user code and may not work for every scenario!
+
+First of all, what contexts are affected and how:
+
+* Application context
+** Works out of the box, no propagation needed
+* Singleton context
+** Works out of the box, no propagation needed
+* Dependent context
+** By nature of this context, this cannot be propagated
+* Request, session, conversation contexts
+** These can be manually propagated if desired
+
+In order to achieve context propagation you generally need the following steps:
+
+* Obtain collection of contextual instances from current thread
+* In another thread, obtain a reference to given context and activate it
+* Feed this newly activated context the instances you previously obtained
+* Perform your tasks
+* Clean up the context by deactivating it
+
+==== New API methods supporting context propagation
+
+There are several new things in Weld API allowing for this.
+Firstly, all contexts supporting propagation now implement `org.jboss.weld.context.WeldAlterableContext`, an interface extending `javax.enterprise.context.spi.AlterableContext`.
+Methods on `WeldAlterableContext` allow to capture current context state, returning a colletion of all contextual instances, as well as clear and set the context state by feeding it a collection of contextual instances.
+
+[source.JAVA, java]
+-------------------------------------------------------------------------------------------------------------
+public interface WeldAlterableContext extends AlterableContext {
+    default <T> Collection<ContextualInstance<T>> getAllContextualInstances();
+    default <T> void clearAndSet(Collection<ContextualInstance<T>> setOfInstances);
+}
+-------------------------------------------------------------------------------------------------------------
+
+In order to get hold of these contexts, the best approach is to use `WeldManager`, an injectable bean providing some capabilities on top of what `BeanManager` has.
+From there, following methods are handy for context propagation:
```

Comment:
I don't like the "From there" much, could it be possibly reworded? Something like "Following WeldManager methods are.."